cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -239,12 +241,60 @@ private void isolate(String node, Set<String> addedAlready) {
     }
 
     Set<String> remainingSources = Sets.intersection(remainingNodes, possibleNewSources);
+    Set<String> processedNodes = new HashSet<>();
+    // Since there can be remaining sources from subdags which don't overlap, they should be split as seperate subdags.
+    // For that, we first create all subdags from remaining sources and keep track of accessible nodes for those
```

Comment:
it would help to draw a diagram in the comments to illustrate what is happening. Otherwise this very hard to describe. You could re-use the one from the test case.